### ISSUE TITLE
[MNG-5835] Maven-Plugin's getLog() ignores -Dorg.slf4j.simpleLogger.defaultLogLevel=warn

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7789]
**[S L](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=thesnoozer)** opened **[MNG-5835](https://issues.apache.org/jira/browse/MNG-5835?redirect=false)** and commented

Hi,
since Maven should supports slf4j-Logging combined with the SLF4J Simple implementation from Maven 3.1.0 onward (http://maven.apache.org/maven-logging.html).
I'm kind of wondering why the default getLog() called from a Plugin ignores the Environment-Variable ``-Dorg.slf4j.simpleLogger.defaultLogLevel=warn``

I'm currently using:
Apache Maven 3.2.3 (33f8c3e1027c3ddde99d3cdebad2656a31e8fdf4; 2014-08-11T22:58:10+02:00)
Maven home: /usr/share/maven-3.2.3
Java version: 1.7.0_80, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-7-oracle/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "linux", version: "3.16.0-33-generic", arch: "amd64", family: "unix"

Tested with different Maven-Versions and different maven-plugin-api Versions, still no success.
Any help is highly appreciated.
Thanks,

PS: Hopefully I can attach my Example-Project which can be executed by using:
mvn clean install && mvn clean package -Pdemo -Dorg.slf4j.simpleLogger.defaultLogLevel=warn

---

**Affects:** 3.2.3, 3.3.3

**Attachments:**
- [hello-maven-plugin.zip](https://issues.apache.org/jira/secure/attachment/12736830/hello-maven-plugin.zip) (_1.55 kB_)
- [hello-maven-plugin2.zip](https://issues.apache.org/jira/secure/attachment/12738261/hello-maven-plugin2.zip) (_1.95 kB_)

**Issue Links:**
- [MNG-6296](https://issues.apache.org/jira/browse/MNG-6296) New option -Dstyle.color is not working
 (_**"is part of"**_)

1 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Joseph Walton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joe@kafsemo.org)** commented

`MavenCli` considers the `-D` arguments passed to Maven after the SLF4J logger has already been initialised:

```
logging( cliRequest );
version( cliRequest );
properties( cliRequest );
```

To work around this, pass the arguments using `MAVEN_OPTS` instead, so they're applied directly by `java` rather than as part of Maven's initialisation:

```
MAVEN_OPTS=-Dorg.slf4j.simpleLogger.defaultLogLevel=warn mvn clean package -Pdemo
```

Alternatively, use `mvn --quiet` which will also set the underlying SLF4J level to `ERROR`.

**Comment by jira-importer** [comment]
**[S L](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=thesnoozer)** commented

Thanks [Joseph Walton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joe@kafsemo.org) for taking your time and suggesting a valid workaround for the example.
However, I just created a slightly changed hello-maven-plugin where even the MAVEN_OPTS are ignored. Also the -Dorg.slf4j.simpleLogger Options are ignored.
This somehow even bypasses the configuration with MAVEN_OPTS.

hello-maven-plugin.zip

```
getLog().info("Info-Message: Hello, world.");
```

hello-maven-plugin2.zip

```
private final Log logger = getLog();
public void execute() {
    logger.debug("Debug-Message: Hello, world.");
 }
```

and now even debug gets displayed:

> MAVEN_OPTS=-Dorg.slf4j.simpleLogger.defaultLogLevel=warn mvn clean package -Pdemo
> [error] ##################################
> [debug] Debug-Message: Hello, world.
> [info] Info-Message: Hello, world.
> [warn] Warn-Message: Hello, world.
> [error] Error-Message: Hello, world.

**Comment by jira-importer** [comment]
**[Joseph Walton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joe@kafsemo.org)** commented

> However, I just created a slightly changed hello-maven-plugin

Don't do that ;-)

If you call `getLog()` in your constructor then you'll [get an instance of `SystemStreamLog`](https://github.com/apache/maven/blob/40d5087b6b134842e2b61a567dbb4bfbcfab7ae6/maven-plugin-api/src/main/java/org/apache/maven/plugin/AbstractMojo.java#L177) that unconditionally logs to `System.out`. Your mojo will be wired with an SLF4J-backed logger **after** construction.

Call `getLog()` during the `execute()` method, and don't cache the value.

If any of this behaviour seems like it could be improved, consider being specific about your use case and your expectations. Otherwise these sound like cases where the documentation should be clearer, rather than the behaviour changed.

**Comment by jira-importer** [comment]
**[S L](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=thesnoozer)** commented

Hi [Joseph Walton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=joe@kafsemo.org) thanks for clarifying the difference behind the scenes. This really helps and I highly appreciate the time you take to look into my problem :-)
I wasn't aware that you get some sort of different logger if you call the getLog before the construction.

In my option not caching the logger value somewhere can under some condition violate the OOP principle of encapsulation.
I personally prefer some sort of LoggerBridge so in case something with the logging change I just need to adjust the Bridge and not my whole code [maybe this is old-fashioned but I suffered once and don't want to repeat that ;)].
I guess this may be some sort of personal problem and to be honest I don't want to have changed anything in the behavior you outlined in your previous post.

Nevertheless, and regrading the workaround, I think that the original problem needs to be addressed. As you outlined this either needs to be addressed in the docs or in the code. From my personal point of view I slightly prefer that issue to be fixed in the code.
This is due to the fact that the SimpleLogger checks whether there are properties present for slf4j logger configuration

> http://grepcode.com/file/repo1.maven.org/maven2/org.slf4j/slf4j-simple/1.7.12/org/slf4j/impl/SimpleLogger.java/#SimpleLogger.init%28%29
>     
>     static void  init() {
>         String defaultLogLevelString = getStringProperty(DEFAULT_LOG_LEVEL_KEY, null);
>         if (defaultLogLevelString != null)
>             DEFAULT_LOG_LEVEL = stringToLevel(defaultLogLevelString);
> 
> whereas getStringProperty should extracts the Property from the System Properties:
> http://grepcode.com/file/repo1.maven.org/maven2/org.slf4j/slf4j-simple/1.7.12/org/slf4j/impl/SimpleLogger.java/#SimpleLogger.getStringProperty%28java.lang.String%29
> ....
> System.getProperty(name);

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This has been fixed with commit b2afafe5a252d9f86b3e43cce8a06dc9cdc7a6d7(MNG-6296).
